shoop/core/telemetry.py

```diff
@@ -0,0 +1,196 @@
+
+
+def set_opt_out(flag):
+    """
+    Set whether this installation is opted-out from telemetry submissions.
+
+    :param flag: Opt-out flag. True for opt-out, false for opt-in (default)
+    :type flag: bool
+    :return: New flag state
+    :rtype: bool
+    """
+    if flag and not is_opt_out():
+        PersistentCacheEntry.objects.create(data=True, **OPT_OUT_KWARGS)
+        return True
+    else:
+        PersistentCacheEntry.objects.filter(**OPT_OUT_KWARGS).delete()
+        return False
+
+
+def get_last_submission_time():
+    try:
+        return PersistentCacheEntry.objects.get(**LAST_DATA_KWARGS).time
+    except ObjectDoesNotExist:
+        return None
+
+
+def get_last_submission_data():
+    try:
+        return safe_json(PersistentCacheEntry.objects.get(**LAST_DATA_KWARGS).data, indent=4)
+    except ObjectDoesNotExist:
+        return None
+
+
+def save_telemetry_submission(data):
+    """
+    Save a blob of data as the latest telemetry submission.
+
+    Naturally updates the latest submission time.
+
+    :param data: A blob of data.
+    :type data: dict
+    """
+    pce, _ = PersistentCacheEntry.objects.get_or_create(defaults={"data": None}, **LAST_DATA_KWARGS)
+    pce.data = data
+    pce.save()
+
+
+def get_telemetry_data(request, indent=None):
+    """
+    Get the telemetry data that would be sent.
+
+    :param request: HTTP request. Optional.
+    :type request: django.http.HttpRequest|None
+    :return: Data blob.
+    :rtype: str
+    """
+    data_dict = {
+        "apps": settings.INSTALLED_APPS,
+        "host": (request.get_host() if request else None),
+        "key": get_installation_key(),
+        "machine": platform.machine(),
+        "platform": platform.platform(),
+        "python_version": sys.version,
+        "shoop_version": shoop.__version__,
+    }
+    return safe_json(data_dict, indent)
+
+
+class TelemetryNotSent(Exception):
+    def __init__(self, message, code):
+        self.message = message
+        self.code = code
+        super(TelemetryNotSent, self).__init__(message, code)
+
+
+def _send_telemetry(request, max_age_hours):
+    if not is_telemetry_enabled():
+        raise TelemetryNotSent("Telemetry not enabled", "disabled")
+
+    if is_opt_out():
+        raise TelemetryNotSent("Telemetry is opted-out", "optout")
+
+    if is_in_grace_period():
+        raise TelemetryNotSent("Telemetry in grace period", "grace")
+
+    if max_age_hours is not None:
+        last_send_time = get_last_submission_time()
+        if last_send_time:
+            if (now() - last_send_time).total_seconds() <= max_age_hours * 60 * 60:
+                raise TelemetryNotSent("Trying to resend too soon", "age")
+
+    data = get_telemetry_data(request)
+    try:
+        resp = requests.post(url=settings.SHOOP_TELEMETRY_URL, data=data)
+    except Exception as exc:
+        data = {
+            "data": data,
+            "error": force_text(exc),
+        }
```

Comment:
The exceptions are handled nicely here, but I'm still a bit worried, since it could block for a long time if the server is unreachable for some reason (no net, no DNS-server, outgoing traffic is firewalled, our end not listening, etc.).

Can we easily make this async, e.g. by doing the stuff from lines 160 to 173 in a multiprocessing.Pool.apply_async or something similar? That would of course change the way response is reported back, because it cannot be returned as normal return value then. We could provide a future to it though, so caller could call something like
```python
telemetry_result = send_telemetry(request)
try:
    result = telemetry_result.get(timeout=1)
except multiprocessing.TimeoutError:
    result = None
...
```